### PR TITLE
Add the "allow resume cancelled" patch to 4.6

### PR DESCRIPTION
**Pull request by tehasdf**

### CODE REVIEW

**Comment by tehasdf** [comment]
Example test scenario:
```bash
# prerequisite for installing an agent is putting the ssh key for the machine
# into /etc/cloudify/key.pem, chown cfyuser, chmod 0400

# need jq to get the last exec id
yum install epel-release
yum install jq

# test 1 - mgmtworker-side operation
mkdir bp
cat > bp/bp.yaml <<EOF
tosca_definitions_version: cloudify_dsl_1_3

imports:
  - http://www.getcloudify.org/spec/cloudify/4.6/types.yaml

node_templates:
 x:
  type: cloudify.nodes.Root
  interfaces:
   foo:
    bar:
     implementation: scripts/foo.py
     executor: central_deployment_agent
 y:
  type: cloudify.nodes.Root
  relationships:
   - type: cloudify.relationships.depends_on
     target: x
  interfaces:
   foo:
    bar:
     implementation: scripts/just_log.py
     executor: central_deployment_agent
EOF

mkdir bp/scripts
cat > bp/scripts/foo.py <<EOF
import time
from cloudify import ctx

ctx.logger.info('starting')
time.sleep(60)
ctx.logger.info('done')

EOF

cat > bp/scripts/just_log.py <<EOF 
from cloudify import ctx

ctx.logger.info('just loggin')
EOF



cfy blu up -bb1 bp/bp.yaml
cfy dep cre -bb1 d1


# we can clear the operations table to make output cleaner to read
sudo -upostgres psql cloudify_db -c "delete from operations"

cfy exe sta -dd1 execute_operation -p operation=foo.bar -p run_by_dependency_order=true --timeout 10
EXEC_ID=$(cfy exe li --json | jq -r '.[-1].id')
cfy exe cancel $EXEC_ID
sleep 3
# check that one is in state pending, and one is started
sudo -upostgres psql cloudify_db -c "select name, state from operations where type='RemoteWorkflowTask'"
# allow the task to finish
sleep 60
# check that one is succeeded and one is pending
sudo -upostgres psql cloudify_db -c "select name, state from operations where type='RemoteWorkflowTask'"

cfy exe resume $EXEC_ID
cfy events list --tail $EXEC_ID

# check that both are succeeded
sudo -upostgres psql cloudify_db -c "select name, state from operations where type='RemoteWorkflowTask'"

# check that the task was run once - it was logged only once
grep 'starting$' /var/log/cloudify/mgmtworker/logs/d1.log


# test 2 - agent-side operation
mkdir bp2
cat >bp2/bp.yaml <<EOF
tosca_definitions_version: cloudify_dsl_1_3

imports:
  - http://www.getcloudify.org/spec/cloudify/4.6/types.yaml

node_templates:
 x:
  type: cloudify.nodes.Compute
  properties:
    ip: 127.0.0.1
    agent_config:
     user: centos
     key: /etc/cloudify/key.pem
  interfaces:
   foo:
    bar:
     implementation: scripts/foo.py

 y:
  type: cloudify.nodes.Root
  relationships:
   - type: cloudify.relationships.depends_on
     target: x
  interfaces:
   foo:
    bar:
     implementation: scripts/just_log.py
     executor: central_deployment_agent
EOF
cp bp/scripts bp2/scripts -fr

cfy blu up -bb2 bp2/bp.yaml
cfy dep cre -bb2 d2
cfy exe start -dd2 install


# we can clear the operations table to make output cleaner to read
sudo -upostgres psql cloudify_db -c "delete from operations"

cfy exe sta -dd2 execute_operation -p operation=foo.bar -p run_by_dependency_order=true --timeout 10
EXEC_ID=$(cfy exe li --json | jq -r '.[-1].id')
cfy exe cancel $EXEC_ID
sleep 3
# check that one is in state pending, and one is sent
sudo -upostgres psql cloudify_db -c "select name, state from operations where type='RemoteWorkflowTask'"
# allow the task to finish
sleep 60

cfy exe resume $EXEC_ID
cfy events list --tail $EXEC_ID

# check that both are succeeded
sudo -upostgres psql cloudify_db -c "select name, state from operations where type='RemoteWorkflowTask'"


# we can clear the operations table to make output cleaner to read
sudo -upostgres psql cloudify_db -c "delete from operations"

# check that there are no `response_xxxx` queues left over, but just 3 per agent (service, operation and workflow; 3 for the installed agent and 3 left over from sanity)
rabbitmqctl -n cloudify-manager@localhost   -p rabbitmq_vhost_default_tenant list_queues

# check that it was only logged once
grep starting /home/centos/x_*/work/logs/d2.log
```

File: patch_files/patches/46_resume_cancelled_cloudify_workflows_tasks.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+diff --git a/cloudify/workflows/tasks.py b/cloudify/workflows/tasks.py
+index cf3aedc..80b14f1 100644
+--- a/cloudify/workflows/tasks.py
++++ b/cloudify/workflows/tasks.py
+@@ -421,6 +421,9 @@ class RemoteWorkflowTask(WorkflowTask):
+         # can safely be rerun
+         if state == TASK_SENDING:
+             return
++        if self.cloudify_context['executor'] != 'host_agent' \
```
Comment (tehasdf) [inline patch_files/patches/46_resume_cancelled_cloudify_workflows_tasks.py:9]: changes in this file are specific to this patch and are not in master, unfortunate

those changes have to do with how we aren't going to patch agents, and so for agent (not mgmtworker) operations it still has to be the dispatcher side that updates the operation state

File: patch_files/patches/46_resume_cancelled_cloudify_workflows_workflow_context.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+         client = self._get_client(task)
+-        client.add_handler(handler)
+         client.add_handler(ping_handler)
++        client.add_handler(handler)
+         task.update({
+             'client': client,
+             'handler': handler,
+             'ping_handler': ping_handler
+         })
+-        client.consume_in_thread()
```
Comment (tehasdf) [inline patch_files/patches/46_resume_cancelled_cloudify_workflows_workflow_context.py:17]: this change is not in master, but will be:
moving `consume_in_thread` to AFTER the callback has been added fixes a race condition between receiving the response message, and adding a callback that will handle it

File: patch_files/patches/46_resume_cancelled_cloudify_amqp_client.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+             del self._response_queues[correlation_id]
+ 
+     def process(self, channel, method, properties, body):
+@@ -624,7 +624,8 @@ class CallbackRequestResponseHandler(_RequestResponseHandlerBase):
+             logger.error('Error parsing response: {0}'.format(body))
+             return
+         if properties.correlation_id in self._callbacks:
+-            self._callbacks[properties.correlation_id](response)
++            channel.queue_delete(self.queue)
++            self._callbacks[properties.correlation_id](response, channel)
```
Comment (tehasdf) [inline patch_files/patches/46_resume_cancelled_cloudify_amqp_client.py]: this change isn't needed and will be reverted (but seeing as re-generating a patch is a lot of work, i'll wait for your review first)

File: patch_files/patches/46_resume_cancelled_cloudify_workflows_workflow_context.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
++        client.consume_in_thread()
+         return result
+ 
+     def _set_task_state(self, workflow_task, state, event=None):
+@@ -1328,7 +1328,7 @@ class _TaskDispatcher(object):
+                     state, workflow_task,
+                     events.send_task_event_func_remote, event)
+ 
+-    def _received(self, task_id, client, response):
++    def _received(self, task_id, client, response, channel):
```
Comment (tehasdf) [inline patch_files/patches/46_resume_cancelled_cloudify_workflows_workflow_context.py]: this change isn't needed and will be reverted (it's the other side of the to-be-reverted change in amqp_client.py)

**Comment by geokala** [comment]
Patch applies, and provided test scenarios pass. Approving.

**Review by geokala — APPROVED** [review]

**Pull request merged**